runtime/src/snapshot_utils.rs

```diff
@@ -1538,29 +1535,45 @@ pub fn snapshot_bank(
     bank_snapshots_dir: impl AsRef<Path>,
     snapshot_archives_dir: impl AsRef<Path>,
     snapshot_version: SnapshotVersion,
-    archive_format: &ArchiveFormat,
+    archive_format: ArchiveFormat,
     hash_for_testing: Option<Hash>,
+    snapshot_type: Option<SnapshotType>,
 ) -> Result<()> {
-    let storages = root_bank.get_snapshot_storages(None);
+    let snapshot_storages = snapshot_type.map_or_else(SnapshotStorages::default, |snapshot_type| {
+        let incremental_snapshot_base_slot = match snapshot_type {
+            SnapshotType::IncrementalSnapshot(incremental_snapshot_base_slot) => {
+                Some(incremental_snapshot_base_slot)
+            }
+            _ => None,
+        };
+        root_bank.get_snapshot_storages(incremental_snapshot_base_slot)
+    });
     let mut add_snapshot_time = Measure::start("add-snapshot-ms");
-    add_bank_snapshot(&bank_snapshots_dir, root_bank, &storages, snapshot_version)?;
+    add_bank_snapshot(
+        &bank_snapshots_dir,
+        root_bank,
+        &snapshot_storages,
+        snapshot_version,
+    )?;
     add_snapshot_time.stop();
     inc_new_counter_info!("add-snapshot-ms", add_snapshot_time.as_ms() as usize);
 
     // Package the relevant snapshots
     let highest_bank_snapshot_info = get_highest_bank_snapshot_info(&bank_snapshots_dir)
         .expect("no snapshots found in config bank_snapshots_dir");
+    assert_eq!(highest_bank_snapshot_info.slot, root_bank.slot());
```

Comment:
Added this assert, since it was just assumed before.

Reply:
sanity check here for my understanding 😃. Who's populating the `bank_snapshots_dir` here: https://github.com/solana-labs/solana/blob/6d246031d06b81ee21d591a267a50f9f76d93eac/runtime/src/snapshot_utils.rs#L1562

It can't have gotten a new root while AccountsBackgroundService was calliing `snapshot_bank()`?

Reply:
The SnapshotConfig is what ultimately sets the `bank_snapshots_dir`. The bank snapshots created by `snapshot_utils::add_bank_snapshot()` is what ends up in this directory. I've added this comment, which hopefully describes the issue sufficiently:

```
    // The result of get_highest_bank_snapshot_info() must be the same as the bank snapshot created
    // by add_bank_snapshot() above.  Otherwise, the AccountsPackage would hardlink the wrong bank
    // snapshot relative to the root_bank's slot and snapshot storages.
```

Does that look OK?

Reply:
hmmm I see, does it make sense to just use the result of the `add_bank_snapshot` then instead of throwing away that result and refetching it via `get_highest_bank_snapshot_info()`

Reply:
Yep, love it! I wasn't sure if this behavior as written was originally intended as a check for the filesystem or something, but with the checks now, that's not an issue. Done!